### PR TITLE
Refactor websocket

### PR DESCRIPTION
**Pull request by tofarr**
**Refactor client side web socket control for better maintainability.**

- [] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
So, when implementing the reconnect, I kept hitting walls. Eventually I figured some of the problem was the way we use sockets on the client side - it feels that we are trying to use it like JQuery. This refactor:

* Turns the websocket from a hook in into a component, so that it is added / removed like any other (And removing closes the socket automatically)
* Formalizes the state a socket expose in it's hook (A status enum, an array of events, and a send method)
* Formalizes the inputs to the socket (token, ghToken, settings, and a new enabled flag)
* Separates the socket state from the events the client executes (Initial query, etc.)

This will make reconnect logic far easier to implement, as it can mostly be isolated to `WsClientProvider`.

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:c642be1-nikolaik   --name openhands-app-c642be1   docker.all-hands.dev/all-hands-ai/openhands:c642be1
```

### CODE REVIEW

File: frontend/src/routes/_oh.tsx [hunk 1]
```diff
@@ -313,11 +303,9 @@ export default function MainApp() {
...
               To continue, connect an OpenAI, Anthropic, or other LLM account
             </p>
-            {isConnected && (
-              <p className="text-xs text-danger">
-                Changing settings during an active session will end the session
-              </p>
-            )}
+            <p className="text-xs text-danger">
+              Changing settings during an active session will end the session
+            </p>
```
Comment (amanape) [inline frontend/src/routes/_oh.tsx:308]: Note that the warning will always be rendered

image

File: frontend/src/utils/verified-models.ts [hunk 2]
```diff
@@ -20,6 +20,7 @@ export const VERIFIED_ANTHROPIC_MODELS = [
   "claude-2",
   "claude-2.1",
   "claude-3-5-sonnet-20240620",
+  "claude-3-5-sonnet-20241022",
```
Comment (amanape) [inline frontend/src/utils/verified-models.ts:23]: If this is to prevent the failing test, I've already fixed it in #4887

File: frontend/src/services/actions.ts [hunk 3]
```diff
@@ -138,22 +141,14 @@ export function handleStatusMessage(message: StatusMessage) {
...
-    handleObservationMessage(socketMessage);
-  } else if ("status_update" in socketMessage) {
-    handleStatusMessage(socketMessage);
+export function handleAssistantMessage(message: Record<string, unknown>) {
+  if (message.action) {
+    handleActionMessage(message as unknown as ActionMessage);
+  } else if (message.observation) {
+    handleObservationMessage(message as unknown as ObservationMessage);
+  } else if (message.status_update) {
+    handleStatusMessage(message as unknown as StatusMessage);
```
Comment (amanape) [inline frontend/src/services/actions.ts:150]: BTW we have some improved (but still unused) types in [frontend/src/types/core](https://github.com/All-Hands-AI/OpenHands/tree/main/frontend/src/types/core). No need to change them yet, just making you aware of it if you aren't already

File: frontend/src/components/event-handler.tsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+export function EventHandler({ children }: EventHandlerProps) {
+  const { events, status, send } = useWsClient();
+  const statusRef = React.useRef<WsClientProviderStatus | null>(null);
+  const runtimeActive = status === WsClientProviderStatus.ACTIVE;
+  const fetcher = useFetcher();
+  const dispatch = useDispatch();
+  const { files, importedProjectZip } = useSelector(
+    (state: RootState) => state.initalQuery,
+  );
+  const { ghToken, repo }: LoaderData = useLoaderData();
```
Comment (amanape) [inline frontend/src/components/event-handler.tsx:69]: After adding `import { clientLoader as appClientLoader } from "#/routes/_oh.app";`:
```suggestion
  const { ghToken, repo } = useLoaderData<typeof appClientLoader>();
```

You can then remove the `LoaderData` interface in preference of the source type

File: frontend/src/components/event-handler.tsx [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        }),
+      );
+      return;
+    }
+    handleAssistantMessage(event);
+  }, [events.length]);
+
+  React.useEffect(() => {
+    if (statusRef.current === status) {
+      return; // This is a check because of strict mode - if the status did not change, don't do anything
```
Comment (amanape) [inline frontend/src/components/event-handler.tsx:123]: We have a `useEffectOnce` util
https://github.com/All-Hands-AI/OpenHands/blob/bf8ccc8fc38b3bce2ef8dea8092082d0c96d5fc3/frontend/src/utils/use-effect-once.ts#L7-L18

File: frontend/src/components/event-handler.tsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+interface LoaderData {
+  settings: Settings;
+  token: string;
+  ghToken: string;
+  repo: string;
+  q: string;
+  lastCommit: string;
+}
+
+export function EventHandler({ children }: EventHandlerProps) {
```
Comment (amanape) [inline frontend/src/components/event-handler.tsx:60]:
```suggestion
export function EventHandler({ children }: React.PropsWithChildren) {
```

File: frontend/src/components/event-handler.tsx
Comment (amanape) [inline frontend/src/components/event-handler.tsx:1]: Nice

**Review by amanape — CHANGES_REQUESTED** [review]

File: frontend/src/context/ws-client-provider.tsx [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+  React.useEffect(() => {
+    let ws = wsRef.current;
+
+    // If disabled close any existing websockets...
+    if (!enabled) {
+      if (ws) {
+        ws.close();
+      }
+      wsRef.current = null;
+      return () => {};
```
[the remark this answers is not in the document]
Comment (tofarr) [reply frontend/src/context/ws-client-provider.tsx:103]: Yeah, but the linter complains if the return type is not consistent.

File: frontend/src/context/ws-client-provider.tsx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    return () => {
+      ws.removeEventListener("open", handleOpen);
+      ws.removeEventListener("message", handleMessage);
+      ws.removeEventListener("error", handleError);
+      ws.removeEventListener("close", handleClose);
+    };
+  }, [enabled, token, ghToken]);
+
+  // Strict mode mounts and unmounts each component twice, so we have to wait in the destructor
+  // before actually closing the socket and cancel the operation if the component gets remounted.
```
[the remark this answers is not in the document]
Comment (tofarr) [reply frontend/src/context/ws-client-provider.tsx:142]: UseEffectOnce will execute the effect once the first time the component mounts, but then not again after this, which means if the tokens or enabled flag changes then the effect will not re-execute.

I thought about adding a custom effect `useEffectIfActuallyChanged`, but decided against it because this is what useEffect is actually meant to do, and it only does not to test whether components actually mount and unmount properly. Our websocket is kind of a corner case where we don't want the mount / unmount behavior.

File: frontend/src/components/event-handler.tsx (hunk 4 above)
Comment (tofarr) [reply frontend/src/components/event-handler.tsx]: Done!

File: frontend/src/context/ws-client-provider.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (amanape) [reply frontend/src/context/ws-client-provider.tsx:142]: I'm referring to the one below the comment which is already executed only once when the component mounts, as implied by the empty dependency array, not the one above 😄

File: frontend/src/components/event-handler.tsx (hunk 5 above)
Comment (amanape) [reply frontend/src/components/event-handler.tsx:110]: Woops! This one does have flags in the dependency array

File: frontend/src/routes/_oh.tsx (hunk 1 above)
Comment (amanape) [reply frontend/src/routes/_oh.tsx:308]: This actually may be annoying. With these changes we're either always going to show this error or never since the parent does not have access to the context. 🤔

**Review by amanape — APPROVED** [review]

File: frontend/src/context/ws-client-provider.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (tofarr) [reply frontend/src/context/ws-client-provider.tsx:142]: Maybe I misunderstood? I was talking about this one:
```
 React.useEffect(() => {
    const timeout = closeRef.current;
    if (timeout != null) {
      clearTimeout(timeout);
    }

    return () => {
      closeRef.current = setTimeout(() => {
        wsRef.current?.close();
      }, 100);
    };
  }, []);
  ```
  
  In this case, if I `useEffectOnce`...
  * `useEffectOnce` does not return a destructor - the destructor is what I am primarily interested in here - it is where I close the websocket if it is still open
  * Even if it did return a destructor, when should it be invoked? It would need timeout logic like I have - otherwise the destructor would always be invoked the first time the component is unmounted, closing the websocket.
